enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,654 @@
+and field for headers to be set/replaced/appended/deleted.
+
+```go
+// IngressControllerSpec is the specification of the desired behavior of the
+// IngressController.
+type IngressControllerSpec struct {
+    // httpHeaders defines policy for HTTP headers.
+    //
+    // If this field is empty, the default values are used.
+    //
+    // +optional
+    HTTPHeaders *IngressControllerHTTPHeaders `json:"httpHeaders,omitempty"`
+}
+```
+
+```go
+
+type IngressControllerHTTPHeaders struct {
+    // headersManipulation specifies options for set/replace/append/delete headers.
+    // +union
+	HeadersManipulation IngressControllerHeadersManipulation `json:"headersManipulation,omitempty"`
+}
+
+```
+
+```go
+// IngressControllerHeadersManipulation specifies options for set/replace/append/delete headers.
+type IngressControllerHeadersManipulation struct {
+    // action lets you specify headers for Set or Delete.
+    // +unionDiscriminator
+    // +kubebuilder:validation:Enum:=Set;Delete
+    // +kubebuilder:validation:Required
+    Action IngressControllerSetHTTPHeaderAction `json:"action,omitempty"`
+
+    // Set defines HTTP headers that should be set or replaced or appended.
+    // If this field is empty, no headers are added.
+    //
+    // Note that this option only applies to cleartext HTTP connections
+    // and to secure HTTP connections for which the ingress controller
+    // terminates encryption (that is, edge-terminated or reencrypt
+    // connections).  Headers cannot be set(replace)/append for TLS passthrough
+    // connections.
+    //
+    // +kubebuilder:validation:Optional
+    // +optional
+    Set *IngressControllerSetHTTPHeaders `json:"set,omitempty"`
+
+    // delete defines HTTP headers that should be deleted.
+    // If this field is empty, no headers are deleted.
+    //
+    // Note that this option only applies to cleartext HTTP connections
+    // and to secure HTTP connections for which the ingress controller
+    // terminates encryption (that is, edge-terminated or reencrypt
+    // connections).  Headers cannot be set(replace)/append for TLS passthrough
+    // connections.
+    //
+    // +kubebuilder:validation:Optional
+    // +optional
+    Delete *IngressControllerDeleteHTTPHeaders `json:"delete,omitempty"`
+}
+```
+
+The IngressControllerSetHTTPHeaders type has a Response field,
+of type []IngressControllerSetHTTPHeader, for specifying headers to set/append:
+
+```go
+// The IngressControllerSetHTTPHeaders type has a Response field,
+// type []IngressControllerSetHTTPHeader, for specifying headers to set(replace) or append:
+
+type IngressControllerSetHTTPHeaders struct {
+    // response specifies which HTTP response headers to set(replace) or append.
+    //
+    // If this field is empty, no response headers are set(replace) or append.
+    //
+    // +nullable
+    // +optional
+    Response []IngressControllerSetHTTPHeader `json:"response,omitempty"`
```

Comment:
`Response` is repeated in Set and Delete structures. It can be moved up to `headerManipulation` to reduce the redundancy. Also, it seems to be more fundamental as parameter than set and delete.

Reply:
Done. The naming: `headerManipulation` tag can still be discussed though, maybe a better name will come up into mind.

Reply:
api changes are updated